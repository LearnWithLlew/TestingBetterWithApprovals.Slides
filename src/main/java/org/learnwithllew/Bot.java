package org.learnwithllew;

import org.learnwithllew.helper.ExistingCustomerResponses;
import org.learnwithllew.helper.HardcodedBotResponses;
import org.learnwithllew.helper.ProspectiveCustomerResponses;
import org.learnwithllew.helper.UnknownCustomerResponses;

import java.sql.SQLOutput;
import java.util.*;

public class Bot {
    private final EnumMap<CustomerType, HardcodedBotResponses> responses = new EnumMap<>(CustomerType.class);
    private final BotOutput output;
    private final Map<String, List<String>> receivedMessages = new HashMap<>();

    private CustomerType customerType = CustomerType.UNKNOWN;

    public Bot(BotOutput output) {
        this.output = output;
        responses.put(CustomerType.UNKNOWN, new UnknownCustomerResponses());
        responses.put(CustomerType.EXISTING, new ExistingCustomerResponses());
        responses.put(CustomerType.PROSPECTIVE, new ProspectiveCustomerResponses());
    }

    public void receive(EventNotification event) {
        recordMessage(event);
        List<String> messagesSoFar = receivedMessages.get(event.getConversationId());

        BotAction response = handleEvent(messagesSoFar);
        output.send(event.getConversationId(), response);
        rememberCustomerType(event);
    }

    private BotAction handleEvent(List<String> messagesSoFar) {
        HardcodedBotResponses botResponses = getHardcodedResponses(customerType);
        BotAction response = botResponses.search(messagesSoFar);
        if (response == null) {
            return new BotAction(List.of(new SendPlainMessageCommand("Hm, I'm not sure I can help you with this.\nLet me transfer to an operator...")));
        }
        return response;
    }

    private HardcodedBotResponses getHardcodedResponses(CustomerType customerType) {
        return responses.get(customerType);
    }

    private void recordMessage(EventNotification event) {
        receivedMessages.computeIfAbsent(event.getConversationId(), it -> new ArrayList<>());
        List<String> messagesSoFar = receivedMessages.get(event.getConversationId());
        event.getEvents().forEach(m -> messagesSoFar.add(m.message()));
    }

    private void rememberCustomerType(EventNotification event) {
        boolean isCustomerResponse = event.getEvents().stream().anyMatch(it -> it.message().contains("Yes, I'm a customer"));
        if (isCustomerResponse) {
            customerType = CustomerType.EXISTING;
        }

        boolean isNoResponse = event.getEvents().stream().anyMatch(it -> it.message().equals("no") || it.message().equals("No, I'm not"));
        if (isNoResponse) {
            customerType = CustomerType.PROSPECTIVE;
        }
    }
}
