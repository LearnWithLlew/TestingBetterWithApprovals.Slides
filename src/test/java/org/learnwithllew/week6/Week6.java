package org.learnwithllew.week6;

import org.apache.commons.lang3.StringUtils;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;
import org.learnwithllew.Bot;
import org.learnwithllew.BotOutput;
import org.learnwithllew.week3.StoryBoard;
import org.learnwithllew.week5.Conversation;

import java.util.ArrayList;
import java.util.List;

import static org.lambda.query.Query.select;

// when everything is done
public class Week6 {

    @Test
    void testConversations() {
        var conversations = List.of(
            conversation("hi"),
            conversation("hi", "hi"),
            conversation("hi", "hi", "hi"),
            conversation("hi", "pay bill"),
            conversation("hi", "pay bill", "Yes, I'm a customer"),
            conversation("hi", "pay bill", "no"),
            conversation("hi", "talk to an operator", "Yes, I'm a customer"),
            conversation("oh hi there, how are you doing"),
            conversation("pay bill"),
            conversation("pay bill", "I like swimming"),
            conversation("pay bill", "I like coffee", "I like tea", "no"),
            conversation("walk my dog", "Yes, I'm a customer"),
            conversation("talk to an operator"),
            conversation("pay bill", "Yes, I'm a customer")
                .thenConversation("pay bill"),
            conversation("pay bill", "No, I'm not")
                .thenConversation("pay bill")
        );

        Approvals.verifyAll("Chatbot conversations", conversations, c -> haveConversations(c));
    }

    private String haveConversations(List<Conversation> conversations) {
        BotOutput output = new BotOutput();
        Bot bot = new Bot(output);

        var separator = "****************************************************************\n";
        String conversationText = select(conversations, Conversation::printMessages).join(", ");
        var storyBoard = "%s* %s%s\n%s".formatted(separator, conversationText,
            StringUtils.leftPad("*", separator.length() - 3 - conversationText.length()), separator);
        for (int i = 0; i < conversations.size(); i++) {
            var conversation = conversations.get(i);
            if (1 < conversations.size()) {
                storyBoard += String.format("%s***** Conversation %s *****\n", i == 0 ? "" : "\n", i + 1);
            }
            storyBoard += StoryBoard.create(bot, output, conversation.messages);

        }
        return storyBoard;
    }

    private MultipleConversations conversation(String... messages) {
        return new MultipleConversations(new Conversation(messages));
    }

    private class MultipleConversations extends ArrayList<Conversation> {
        public MultipleConversations(Conversation conversations) {
            add(conversations);
        }

        public MultipleConversations thenConversation(String... messages) {
            add(new Conversation(messages));
            return this;
        }
    }
}
