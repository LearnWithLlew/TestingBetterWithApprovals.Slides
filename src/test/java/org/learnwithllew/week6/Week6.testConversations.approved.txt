Chatbot conversations


****************************************************************
* [hi]                                                         *
****************************************************************
[Customer]: hi
[     Bot]: Hi there! I'm your virtual assistant.
[     Bot]: What would you like to do today?

****************************************************************
* [hi, hi]                                                     *
****************************************************************
[Customer]: hi
[     Bot]: Hi there! I'm your virtual assistant.
[     Bot]: What would you like to do today?
[Customer]: hi
[     Bot]: Hmmm, tell me a little more so I can help you.
            What would you like to do today?

****************************************************************
* [hi, hi, hi]                                                 *
****************************************************************
[Customer]: hi
[     Bot]: Hi there! I'm your virtual assistant.
[     Bot]: What would you like to do today?
[Customer]: hi
[     Bot]: Hmmm, tell me a little more so I can help you.
            What would you like to do today?
[Customer]: hi
[     Bot]: Let me try to help you.
[     Bot]: Are you a customer?
[        ]:   1) Yes, I'm a customer
[        ]:   2) No, I'm not

****************************************************************
* [hi, pay bill]                                               *
****************************************************************
[Customer]: hi
[     Bot]: Hi there! I'm your virtual assistant.
[     Bot]: What would you like to do today?
[Customer]: pay bill
[     Bot]: Let me try to help you.
[     Bot]: Are you a customer?
[        ]:   1) Yes, I'm a customer
[        ]:   2) No, I'm not

****************************************************************
* [hi, pay bill, Yes, I'm a customer]                          *
****************************************************************
[Customer]: hi
[     Bot]: Hi there! I'm your virtual assistant.
[     Bot]: What would you like to do today?
[Customer]: pay bill
[     Bot]: Let me try to help you.
[     Bot]: Are you a customer?
[        ]:   1) Yes, I'm a customer
[        ]:   2) No, I'm not
[Customer]: Yes, I'm a customer
[     Bot]: transfers to 'self_service'

****************************************************************
* [hi, pay bill, no]                                           *
****************************************************************
[Customer]: hi
[     Bot]: Hi there! I'm your virtual assistant.
[     Bot]: What would you like to do today?
[Customer]: pay bill
[     Bot]: Let me try to help you.
[     Bot]: Are you a customer?
[        ]:   1) Yes, I'm a customer
[        ]:   2) No, I'm not
[Customer]: no
[     Bot]: transfers to 'operator'

****************************************************************
* [hi, talk to an operator, Yes, I'm a customer]               *
****************************************************************
[Customer]: hi
[     Bot]: Hi there! I'm your virtual assistant.
[     Bot]: What would you like to do today?
[Customer]: talk to an operator
[     Bot]: Let me try to help you.
[     Bot]: Are you a customer?
[        ]:   1) Yes, I'm a customer
[        ]:   2) No, I'm not
[Customer]: Yes, I'm a customer
[     Bot]: transfers to 'operator'

****************************************************************
* [oh hi there, how are you doing]                             *
****************************************************************
[Customer]: oh hi there, how are you doing
[     Bot]: Hi there! I'm your virtual assistant.
[     Bot]: Let me try to help you.
[     Bot]: Are you a customer?
[        ]:   1) Yes, I'm a customer
[        ]:   2) No, I'm not

****************************************************************
* [pay bill]                                                   *
****************************************************************
[Customer]: pay bill
[     Bot]: Hi there! I'm your virtual assistant.
[     Bot]: Let me try to help you.
[     Bot]: Are you a customer?
[        ]:   1) Yes, I'm a customer
[        ]:   2) No, I'm not

****************************************************************
* [pay bill, I like swimming]                                  *
****************************************************************
[Customer]: pay bill
[     Bot]: Hi there! I'm your virtual assistant.
[     Bot]: Let me try to help you.
[     Bot]: Are you a customer?
[        ]:   1) Yes, I'm a customer
[        ]:   2) No, I'm not
[Customer]: I like swimming
[     Bot]: I'd be happy to help you with this, But first, I need to know: are you a customer?
[        ]:   1) Yes, I'm a customer
[        ]:   2) No, I'm not

****************************************************************
* [pay bill, I like coffee, I like tea, no]                    *
****************************************************************
[Customer]: pay bill
[     Bot]: Hi there! I'm your virtual assistant.
[     Bot]: Let me try to help you.
[     Bot]: Are you a customer?
[        ]:   1) Yes, I'm a customer
[        ]:   2) No, I'm not
[Customer]: I like coffee
[     Bot]: I'd be happy to help you with this, But first, I need to know: are you a customer?
[        ]:   1) Yes, I'm a customer
[        ]:   2) No, I'm not
[Customer]: I like tea
[     Bot]: I'd be happy to help you with this, But first, I need to know: are you a customer?
[        ]:   1) Yes, I'm a customer
[        ]:   2) No, I'm not
[Customer]: no
[     Bot]: transfers to 'operator'

****************************************************************
* [walk my dog, Yes, I'm a customer]                           *
****************************************************************
[Customer]: walk my dog
[     Bot]: Hi there! I'm your virtual assistant.
[     Bot]: Let me try to help you.
[     Bot]: Are you a customer?
[        ]:   1) Yes, I'm a customer
[        ]:   2) No, I'm not
[Customer]: Yes, I'm a customer
[     Bot]: transfers to 'operator'

****************************************************************
* [talk to an operator]                                        *
****************************************************************
[Customer]: talk to an operator
[     Bot]: Hi there! I'm your virtual assistant.
[     Bot]: Let me try to help you.
[     Bot]: Are you a customer?
[        ]:   1) Yes, I'm a customer
[        ]:   2) No, I'm not

****************************************************************
* [pay bill, Yes, I'm a customer], [pay bill]                  *
****************************************************************
***** Conversation 1 *****
[Customer]: pay bill
[     Bot]: Hi there! I'm your virtual assistant.
[     Bot]: Let me try to help you.
[     Bot]: Are you a customer?
[        ]:   1) Yes, I'm a customer
[        ]:   2) No, I'm not
[Customer]: Yes, I'm a customer
[     Bot]: transfers to 'self_service'

***** Conversation 2 *****
[Customer]: pay bill
[     Bot]: Hi there! I'm your virtual assistant.
[     Bot]: Let me try to help you.
[     Bot]: transfers to 'self_service'

****************************************************************
* [pay bill, No, I'm not], [pay bill]                          *
****************************************************************
***** Conversation 1 *****
[Customer]: pay bill
[     Bot]: Hi there! I'm your virtual assistant.
[     Bot]: Let me try to help you.
[     Bot]: Are you a customer?
[        ]:   1) Yes, I'm a customer
[        ]:   2) No, I'm not
[Customer]: No, I'm not
[     Bot]: transfers to 'operator'

***** Conversation 2 *****
[Customer]: pay bill
[     Bot]: Let me try to help you.
[     Bot]: transfers to 'operator'

